### PR TITLE
feat(wh): metadata passed in webhooks

### PR DESCRIPTION
**Pull request by jonahkaye**
Refs: #1018

### Dependencies

- Downstream: https://github.com/metriport/metriport/pull/1106

### Description

- v1 version of https://github.com/metriport/metriport/pull/1069
- We update the POST /document/query & POST /consolidated to accept an additional param (body) containing the request’s metadata (dictionary)
- When the request entity is updated to "completed"/"failed" from "processing", we return the metadata along with the WH
- The data included in the body can be any json within the metadata field, i.e.
```
{
    "metadata": 
    {
        "externalRequestId": "xxx-xxx-xxx"
    }  
}
```
but not 
```
{
        "externalRequestId": "xxx-xxx-xxx"
}
```

### Release Plan

- [ ] merge
- [ ] test on staging
- [ ] NPM release
- [ ] release

### CODE REVIEW

File: packages/core/src/external/aws/lambda-logic/bundle-to-html.ts [hunk 1]
```diff
@@ -585,12 +585,7 @@ function createDiagnosticReportsSection(
       const hasNoReasonForVisit = !reasonForVisit || reasonForVisit?.length === 0;
       const hasNoDocumentation = !filteredDocumentation || filteredDocumentation?.length === 0;
 
-      if (
-        hasNoLabs &&
-        hasNoProgressNotes &&
-        hasNoReasonForVisit &&
-        hasNoDocumentation
-      ) {
+      if (hasNoLabs && hasNoProgressNotes && hasNoReasonForVisit && hasNoDocumentation) {
```
Comment (jonahkaye) [inline packages/core/src/external/aws/lambda-logic/bundle-to-html.ts:588]: linting changes

**Review by jonahkaye — COMMENTED** [review]
Reviewed changes!

File: packages/api/src/external/commonwell/document/document-query-sandbox.ts [hunk 2]
```diff
@@ -35,12 +34,12 @@ export async function sandboxGetDocRefsAndUpsert({
   organization,
   patient,
   requestId,
+  cxPayload,
 }: {
   organization: Organization;
   patient: Patient;
-  facility: Facility;
```
Comment (jonahkaye) [inline packages/api/src/external/commonwell/document/document-query-sandbox.ts:41]: removed since was unused variable causing linting issue

File: packages/api/src/external/commonwell/document/document-query-sandbox.ts [hunk 3]
```diff
@@ -5,7 +5,6 @@ import {
   processPatientDocumentRequest,
 } from "../../../command/medical/document/document-webhook";
 import { appendDocQueryProgress } from "../../../command/medical/patient/append-doc-query-progress";
-import { Facility } from "../../../domain/medical/facility";
```
Comment (jonahkaye) [inline packages/api/src/external/commonwell/document/document-query-sandbox.ts:8]: removed since was unused import

File: packages/api/src/external/commonwell/document/document-query-sandbox.ts [hunk 4]
```diff
@@ -138,7 +137,7 @@ export async function sandboxGetDocRefsAndUpsert({
       log(`Skipping file ${fileTitle} as it already exists`);
       const isDocConvertible = isConvertible(entry.content?.mimeType);
 
-      if  (isDocConvertible) {
```
Comment (jonahkaye) [inline packages/api/src/external/commonwell/document/document-query-sandbox.ts:141]: linting

File: packages/api/src/external/commonwell/document/document-query-sandbox.ts [hunk 5]
```diff
@@ -153,13 +152,16 @@ export async function sandboxGetDocRefsAndUpsert({
       status: "completed",
       successful: entries.length,
     },
-    convertProgress: docsToConvert <= 0 ? {
```
Comment (jonahkaye) [inline packages/api/src/external/commonwell/document/document-query-sandbox.ts:156]: linting

**Review by jonahkaye — COMMENTED** [review]
reviewed post merge to develop

File: packages/api/src/command/medical/document/document-query.ts [hunk 6]
```diff
@@ -41,12 +41,14 @@ export async function queryDocumentsAcrossHIEs({
   patientId,
   facilityId,
   override,
+  cxPayload,
```
Comment (leite08) [inline packages/api/src/command/medical/document/document-query.ts:44]: I'd call this `cxRequestMeta`  or `cxRequestMetadata` as payload doesn't indicate much what that's about. Here and everywhere else.

File: packages/api/src/command/medical/document/document-webhook.ts [hunk 7]
```diff
@@ -39,13 +39,14 @@ export const processPatientDocumentRequest = async (
...
-  documents?: DocumentReferenceDTO[]
+  documents?: DocumentReferenceDTO[],
+  cxPayload?: object
 ): Promise<void> => {
   try {
     const settings = await getSettingsOrFail({ id: cxId });
     // create a representation of this request and store on the DB
     const payload: WebhookPatientDataPayloadWithoutMessageId = {
-      patients: [{ patientId, documents, status }],
+      patients: [{ patientId, documents, status, cxPayload }],
```
Comment (leite08) [inline packages/api/src/command/medical/document/document-webhook.ts:49]: I think this request meta should be a top-level object. Just gotta figure out a good name, since we already use `meta` for our own metadata. Maybe something like `documentQueryMeta`?

File: packages/api/src/command/medical/patient/consolidated-webhook.ts [hunk 8]
```diff
@@ -54,7 +57,7 @@ export const processConsolidatedDataWebhook = async ({
 
     // create a representation of this request and store on the DB
     const payload: PayloadWithoutMeta = {
-      patients: [{ patientId, status, bundle, filters }],
+      patients: [{ patientId, status, bundle, filters, cxPayload }],
```
Comment (leite08) [inline packages/api/src/command/medical/patient/consolidated-webhook.ts:60]: ditto, move to top-level if we move as above

File: packages/api/src/command/medical/patient/consolidated-webhook.ts [hunk 9]
```diff
@@ -41,11 +42,13 @@ export const processConsolidatedDataWebhook = async ({
   status,
   bundle,
   filters,
+  cxPayload,
 }: {
   patient: Pick<Patient, "id" | "cxId">;
   status: ConsolidatedWebhookStatus;
   bundle?: Bundle<Resource>;
   filters?: Filters;
+  cxPayload?: object;
```
Comment (leite08) [inline packages/api/src/command/medical/patient/consolidated-webhook.ts:51]: shouldn't this be required?

File: packages/api/src/command/medical/document/document-webhook.ts [hunk 10]
```diff
@@ -39,13 +39,14 @@ export const processPatientDocumentRequest = async (
   patientId: string,
   whType: MAPIWebhookType,
   status: MAPIWebhookStatus,
-  documents?: DocumentReferenceDTO[]
+  documents?: DocumentReferenceDTO[],
+  cxPayload?: object
```
Comment (leite08) [inline packages/api/src/command/medical/document/document-webhook.ts:43]: shouldn't this be required?

**Review by leite08 — CHANGES_REQUESTED** [review]
A few questions/suggestions:
- where do we store the request meta? Maybe use `initPatientDocQuery()`?
- also need to get that ^ from the patient record/DB and pass it to `processPatientDocumentRequest` on `POST /conversion-status`
- missing the update on the docs to represent this new capability

File: packages/api/src/command/medical/patient/consolidated-webhook.ts (hunk 9 above)
Comment (jonahkaye) [reply packages/api/src/command/medical/patient/consolidated-webhook.ts:51]: No because the payload is optional for the customer right?

File: packages/api/src/command/medical/document/document-webhook.ts (hunk 10 above)
Comment (jonahkaye) [reply packages/api/src/command/medical/document/document-webhook.ts:43]: No because the payload is optional for the /document/query endpoint right?

File: packages/api/src/domain/webhook.ts [hunk 11]
```diff
@@ -17,3 +18,5 @@ export const mapiWebhookType = [
 export type MAPIWebhookType = (typeof mapiWebhookType)[number];
 
 export type WebhookType = DAPIWebhookType | MAPIWebhookType;
+
+export const cxWebhookPayloadSchemaOptional = z.record(z.string()).optional();
```
[the remark this answers is not in the document]
Comment (jonahkaye) [reply packages/api/src/domain/webhook.ts:22]: unknown as opposed to object?

**Comment by jonahkaye** [comment]
@leite08 for the place to store the requestMetadata in db, I thought would make most sense to modify updateDocQuery in the /document/query flow and to modify updateConsolidatedQueryProgress in the /consolidated/query flow.

File: packages/api/src/command/medical/document/document-webhook.ts (hunk 10 above)
Comment (leite08) [reply packages/api/src/command/medical/document/document-webhook.ts]: 💯

File: docs/medical-api/more-info/webhooks.mdx [hunk 12]
```diff
@@ -96,6 +96,16 @@ Example payload:
       ...
     }
   ]
+  "cxDocumentRequestMetadata": {
```
Comment (jonahkaye) [inline docs/medical-api/more-info/webhooks.mdx:99]: I added both in the docs to this sample wh payload but it isn't actually the case that you could get both payloads back in one wh.

File: packages/api/src/command/medical/document/document-query.ts [hunk 13]
```diff
@@ -88,6 +90,7 @@ export async function queryDocumentsAcrossHIEs({
     downloadProgress: { status: "processing" },
     requestId,
     reset: true,
+    cxDocumentRequestMetadata,
```
Comment (jonahkaye) [inline packages/api/src/command/medical/document/document-query.ts:93]: stuck with updating db here. optional flag makes it compatible. I think this is the simplest approach as opposed to creating a new dedicated fxn to update the cxDocumentRequestMetadata

File: packages/api/src/command/medical/patient/append-consolidated-query-progress.ts [hunk 14]
```diff
@@ -20,6 +21,7 @@ export async function updateConsolidatedQueryProgress({
   patient,
   progress,
   reset,
+  cxConsolidatedRequestMetadata,
```
Comment (jonahkaye) [inline packages/api/src/command/medical/patient/append-consolidated-query-progress.ts:24]: Updating db here with optional cxConsolidatedRequestMetadata

File: packages/api/src/domain/medical/patient.ts [hunk 15]
```diff
@@ -68,6 +68,8 @@ export type PatientData = {
   documentQueryProgress?: DocumentQueryProgress;
   consolidatedQuery?: QueryProgress;
   externalData?: PatientExternalData;
+  cxDocumentRequestMetadata?: Record<string, string>;
```
Comment (jonahkaye) [inline packages/api/src/domain/medical/patient.ts:71]: two different data types, one for each route

File: packages/api/src/routes/medical/schemas/request-metadata.ts [hunk 16]
```diff
@@ -0,0 +1,5 @@
+import { z } from "zod";
+
+export const cxRequestMetadataSchema = z.object({
+  meta: z.record(z.string().min(1).max(40), z.string().max(40).optional()).optional(),
```
Comment (jonahkaye) [inline packages/api/src/routes/medical/schemas/request-metadata.ts:4]: optional metadata

**Review by jonahkaye — COMMENTED** [review]
Reviewed PR ready for merge

File: packages/api/src/routes/medical/schemas/request-metadata.ts (hunk 16 above)
Comment (leite08) [reply packages/api/src/routes/medical/schemas/request-metadata.ts:4]: Do we also want the value to be limited to 40 chars? Stripe uses 500 and I suggested 256, just trying to understand the thought process.

**Comment by leite08** [comment]
> for the place to store the requestMetadata in db, I thought would make most sense to modify updateDocQuery in the /document/query flow and 

Discussed offline

File: docs/medical-api/more-info/webhooks.mdx [hunk 17]
```diff
@@ -34,6 +34,10 @@ These are messages you can expect to receive in the following scenarios:
 2. Then, if the downloaded documents contained C-CDA/XML files, when the conversion to FHIR has completed (`type` for a patient in the message
    will be `document-conversion`, and at this point you'll be able to query for [patient consolidated data](#patient-consolidated-data) in FHIR-compliant format).
 
+3. Note: You will not recieve both a cxDocumentRequestMetadata and cxConsolidatedRequestMetadata in a single webhook. Below it is shown that way
+   just to demonstrate the way each would look in their respective situations. cxDocumentRequestMetadata when querrying /document/query and cxConsolidatedRequestMetadata
+   when querrying patient/consolidated/query
```
Comment (leite08) [inline docs/medical-api/more-info/webhooks.mdx:39]: - typo: recieve > receive
- update properties' names
- wrap properties with back-tick (shows up as code)
- I suggest we rewrite this to make it a bit simpler, something like this maybe?
> 3. Note: only one of `documentRequestMetadata` and `consolidatedRequestMetadata` will be present on the webhook payload - they're both displayed on the same payload here for reference only.

File: packages/api/src/command/medical/document/document-query.ts [hunk 18]
```diff
@@ -42,14 +43,14 @@ export async function queryDocumentsAcrossHIEs({
   facilityId,
   override,
   cxDocumentRequestMetadata,
-  skipDocQueryStatusCheck = false,
+  forceQuery = false,
```
Comment (leite08) [inline packages/api/src/command/medical/document/document-query.ts:46]: OOC, this is not part of this PR, why showing up as a change here? Did you rebase your branch from `develop`? It seems like a bad merge, since the previous version already had `forceQuery`.

Not a huge deal, mostly so you know and watch out for those, its fairly easy to introduce bugs when merging.

**Review by Goncharo — CHANGES_REQUESTED** [review]
comments here apply https://github.com/metriport/metriport/pull/1106

the 1106 PR should be based off this one

File: packages/api/src/routes/medical/schemas/request-metadata.ts [hunk 19]
```diff
@@ -1,5 +1,11 @@
 import { z } from "zod";
 
 export const cxRequestMetadataSchema = z.object({
-  meta: z.record(z.string().min(1).max(40), z.string().max(40).optional()).optional(),
+  meta: z
+    .record(z.string().min(1).max(40), z.string().max(500))
+    .optional()
+    .refine(record => (record ? Object.keys(record).length <= 50 : true), {
```
Comment (leite08) [inline packages/api/src/routes/medical/schemas/request-metadata.ts:7]: 🤩

**Review by leite08 — CHANGES_REQUESTED** [review]
Left some comments/asks, also missed updating the SDK so cxs can pass the metadata (and example on Docs)

File: packages/api/src/command/medical/document/document-query.ts (hunk 18 above)
Comment (jonahkaye) [reply packages/api/src/command/medical/document/document-query.ts:46]: I merged with develop since I had conflicts. Is this change a bug?

Comment (leite08) [reply packages/api/src/command/medical/document/document-query.ts:46]: No, like I said this is not a big deal, its mostly so you're aware.

The fact you merged/rebased and this line shows up as changed means that there was some mistake in the merging - this line is already in develop as you can see [here](https://github.com/metriport/metriport/blob/2b0249eb5fc4c97c211bc9616c84583f61e9f3cb/packages/api/src/command/medical/document/document-query.ts#L44).

Mostly sharing this so you're aware for next time. 👍

File: docs/medical-api/api-reference/fhir/consolidated-data-query-post.mdx [hunk 20]
```diff
@@ -63,11 +63,21 @@ This will be returned [via a webook](/medical-api/more-info/webhooks#medical-rec
 
 ## Request Body
 
-<BodyField name="meta" type="object" optional>
+<BodyField name="metadata" type="object" optional>
```
Comment (leite08) [inline docs/medical-api/api-reference/fhir/consolidated-data-query-post.mdx:66]: Could move this to a snipped so we have it written only once - any maintenance is less likely to introduce errors.

File: docs/medical-api/more-info/webhooks.mdx [hunk 21]
```diff
@@ -34,18 +34,18 @@ These are messages you can expect to receive in the following scenarios:
 2. Then, if the downloaded documents contained C-CDA/XML files, when the conversion to FHIR has completed (`type` for a patient in the message
    will be `document-conversion`, and at this point you'll be able to query for [patient consolidated data](#patient-consolidated-data) in FHIR-compliant format).
 
-3. Note: You will not recieve both a cxDocumentRequestMetadata and cxConsolidatedRequestMetadata in a single webhook. Below it is shown that way
-   just to demonstrate the way each would look in their respective situations. cxDocumentRequestMetadata when querrying /document/query and cxConsolidatedRequestMetadata
-   when querrying patient/consolidated/query
-
-Example payload:
+3. Note: only one of `documentRequestMetadata` and `consolidatedRequestMetadata` will be present on the webhook meta.data payload.
```
Comment (leite08) [inline docs/medical-api/more-info/webhooks.mdx:37]: forgot to remove this, we don't have these fields anymore

File: docs/medical-api/more-info/webhooks.mdx [hunk 22]
```diff
@@ -34,18 +34,18 @@ These are messages you can expect to receive in the following scenarios:
...
-Example payload:
+3. Note: only one of `documentRequestMetadata` and `consolidatedRequestMetadata` will be present on the webhook meta.data payload. 
 
 ```json
 {
   "meta": {
     "messageId": "<message-id>",
     "when": "<date-time-in-utc>",
     "type": "medical.document-download"
+    "data": {
```
Comment (leite08) [inline docs/medical-api/more-info/webhooks.mdx:45]: I think we should add a section to this page explaining what `data` means. People can start reading about WHs through this page and they won't have any context as to what this means.

File: packages/api/src/command/webhook/webhook.ts [hunk 23]
```diff
@@ -23,7 +23,12 @@ type WebhookPingPayload = {
   ping: string;
 };
 
-export type WebhookMetadataPayload = { messageId: string; when: string; type: string };
+export type WebhookMetadataPayload = {
+  messageId: string;
+  when: string;
+  type: string;
+  data?: unknown;
```
Comment (leite08) [inline packages/api/src/command/webhook/webhook.ts:30]: Suggestion: can we add a TSDoc for the `WebhookMetadataPayload` type, please? It is growing and now it seems a good time to indicate what each property means - especially `data`.

File: packages/api/src/command/medical/patient/consolidated-webhook.ts [hunk 24]
```diff
@@ -50,8 +51,12 @@ export const processConsolidatedDataWebhook = async ({
   const apiType = Product.medical;
   const { id: patientId, cxId } = patient;
   try {
-    const settings = await getSettingsOrFail({ id: cxId });
+    const [settings, currentPatient] = await Promise.all([
+      getSettingsOrFail({ id: cxId }),
+      getPatientOrFail({ id: patientId, cxId }),
+    ]);
 
+    console.log("CURRENT PATIENT PAYLOAD: ", currentPatient.data.cxConsolidatedRequestMetadata);
```
Comment (leite08) [inline packages/api/src/command/medical/patient/consolidated-webhook.ts:59]: I think this will result in a multi-lined log, right? Let's keep logs to one line to avoid flooding the logs @ CW, please

Comment (leite08) [inline packages/api/src/command/medical/patient/consolidated-webhook.ts:59]: I think this will result in a multi-lined log, right? Let's keep logs to one line to avoid flooding the logs @ CW, please

Comment (jonahkaye) [reply packages/api/src/command/medical/patient/consolidated-webhook.ts:59]: I am just removing it - was a debug statement I forgot to delete

File: packages/utils/src/document/document-query.ts [hunk 25]
```diff
@@ -0,0 +1,57 @@
...
+};
+
+const metadata2 = {
+  metadata: {
+    someKey: "someValue2",
+    anotherKey: "anotherValue2",
+  },
+};
+
+async function main() {
```
Comment (jonahkaye) [inline packages/utils/src/document/document-query.ts:30]: Testing for metadata changes to SDK

File: packages/core/src/external/aws/lambda-logic/bundle-to-html.ts [hunk 26]
```diff
@@ -636,7 +636,6 @@ function buildEncounterSections(
             if (noteIsBlacklisted) {
               continue;
             }
-
```
Comment (jonahkaye) [inline packages/core/src/external/aws/lambda-logic/bundle-to-html.ts:639]: linting change

File: packages/api/src/command/webhook/webhook.ts (hunk 23 above)
Comment (jonahkaye) [reply packages/api/src/command/webhook/webhook.ts:36]: TSDoc added

File: packages/api/src/command/medical/document/document-query.ts [hunk 27]
```diff
@@ -83,13 +86,25 @@ export async function queryDocumentsAcrossHIEs({
...
 
-  const updatedPatient = await updateDocQuery({
+  const updatedPatientProgress: Patient = await updateDocQuery({
     patient: { id: patient.id, cxId: patient.cxId },
     downloadProgress: { status: "processing" },
     requestId,
     reset: true,
   });
 
+  const updatedPatientMetadata = await updatePatient({
```
Comment (leite08) [inline packages/api/src/command/medical/document/document-query.ts:96]: we're sending two updates to patient, including one that overwrites all its `data`, and ended up making `updatePatient` to have query specific logic.

I'm sending a commit with a suggestion of code to simplify it both for doc query and consolidated query.

File: packages/api/src/command/medical/patient/append-consolidated-query-progress.ts [hunk 28]
```diff
@@ -8,6 +8,7 @@ export type SetDocQueryProgress = {
   patient: Pick<Patient, "id" | "cxId">;
   progress: QueryProgress;
   reset?: boolean;
+  cxConsolidatedRequestMetadata?: unknown;
```
Comment (leite08) [inline packages/api/src/command/medical/patient/append-consolidated-query-progress.ts:11]: left-over?

File: packages/api/src/command/medical/patient/append-doc-query-progress.ts [hunk 29]
```diff
@@ -9,6 +9,7 @@ export type SetDocQueryProgress = {
   convertibleDownloadErrors?: number;
   increaseCountConvertible?: number;
   requestId?: string | undefined;
+  cxDocumentRequestMetadata?: unknown;
```
Comment (leite08) [inline packages/api/src/command/medical/patient/append-doc-query-progress.ts:12]: left-over?

File: packages/api/src/command/medical/patient/consolidated-get.ts [hunk 30]
```diff
@@ -45,15 +48,31 @@ export async function startConsolidatedQuery({
     return patient.data.consolidatedQuery;
   }
 
+  const updatedPatient = await updatePatient({
```
Comment (leite08) [inline packages/api/src/command/medical/patient/consolidated-get.ts:51]: ditto: two updates in sequence + suggested new approach on a new commit

**Review by leite08 — CHANGES_REQUESTED** [review]
Left some comments and asks for changes.

Also sent a commit with a suggestion of code to simplify it both for doc query and consolidated query. Please be aware that I haven't tested it.

File: packages/api/src/command/medical/patient/update-patient.ts [hunk 31]
```diff
@@ -29,6 +30,10 @@ export const updatePatient = async (patientUpdate: PatientUpdateCmd): Promise<Pa
       personalIdentifiers: sanitized.personalIdentifiers,
       address: sanitized.address,
       contact: sanitized.contact,
+      cxDocumentRequestMetadata:
+        cxDocumentRequestMetadata !== undefined ? cxDocumentRequestMetadata : undefined,
+      cxConsolidatedRequestMetadata:
+        cxConsolidatedRequestMetadata !== undefined ? cxConsolidatedRequestMetadata : undefined,
```
[the remark this answers is not in the document]
Comment (jonahkaye) [reply packages/api/src/command/medical/patient/update-patient.ts:36]: The idea here @leite08 is to reset the payload to be empty if they don't pass in metadata on successive requests. Otherwise, if they pass in metadata on one webhook request and then nothing on subsequent requests. We can remove that from here but that logic ought to be added to storeQueryInit()

File: docs/medical-api/api-reference/fhir/consolidated-data-query-post.mdx
[the remark this answers is not in the document]
Comment (jonahkaye) [reply docs/medical-api/api-reference/fhir/consolidated-data-query-post.mdx:1]: I added what the metadata on the req body is for to the webhook-request-metada.mdx snippet. 
Added metadata as an argument to the example

File: packages/api-sdk/src/medical/client/metriport.ts [hunk 32]
```diff
@@ -264,16 +264,18 @@ export class MetriportMedicalApi {
...
     patientId: string,
     resources?: readonly ResourceTypeForConsolidation[],
     dateFrom?: string,
     dateTo?: string,
-    conversionType?: string
+    conversionType?: string,
+    metadata?: unknown
   ): Promise<QueryStatus> {
-    const resp = await this.api.post(`${PATIENT_URL}/${patientId}/consolidated/query`, undefined, {
+    const resp = await this.api.post(`${PATIENT_URL}/${patientId}/consolidated/query`, metadata, {
```
[the remark this answers is not in the document]
Comment (jonahkaye) [reply packages/api-sdk/src/medical/client/metriport.ts:278]: I am going to go with the second option. Its what we agreed to before I believe.

File: packages/api/src/command/medical/patient/update-patient.ts (hunk 31 above)
[the remark this answers is not in the document]
Comment (leite08) [reply packages/api/src/command/medical/patient/update-patient.ts:36]: What about the fact `updateQuery` is a generic function and can be called in other use cases/contexts? For example, its used on `PUT /patient/:id` and on `POST /update-all`. These would reset the metadata of a doc query that's in progress, for example.

File: packages/api/src/command/medical/document/document-query.ts [hunk 33]
```diff
@@ -83,13 +86,25 @@ export async function queryDocumentsAcrossHIEs({
   const cwData = externalData as PatientDataCommonwell;
   if (!cwData.patientId) return createQueryResponse("failed");
 
-  const updatedPatient = await updateDocQuery({
+  const updatedPatientProgress: Patient = await updateDocQuery({
```
[the remark this answers is not in the document]
Comment (jonahkaye) [reply packages/api/src/command/medical/document/document-query.ts]: ah ok. I just did it for visibility/documentation but sounds good

File: packages/api/src/routes/medical/schemas/request-metadata.ts [hunk 34]
```diff
@@ -0,0 +1,11 @@
...
+
+export const cxRequestMetadataSchema = z.object({
+  metadata: z
+    .record(z.string().min(1).max(40), z.string().max(500))
+    .optional()
+    .refine(record => (record ? Object.keys(record).length <= 50 : true), {
+      message: "Record has too many properties",
+      path: ["meta"],
+    }),
+});
```
[the remark this answers is not in the document]
Comment (jonahkaye) [reply packages/api/src/routes/medical/schemas/request-metadata.ts:11]: Yes and it works. 
The reason as I understand is there is only one field and it is optional, so an empty object is a valid input. So when the endpoint has an empty body, it passes in an empty object to the schema which is valid.

File: packages/utils/src/document/document-query.ts [hunk 35]
```diff
@@ -0,0 +1,57 @@
...
+dotenv.config();
+
+import { MetriportMedicalApi } from "@metriport/api-sdk";
+import { getEnvVarOrFail } from "@metriport/core/util/env-var";
+
+const apiKey = getEnvVarOrFail("API_KEY");
+const apiUrl = getEnvVarOrFail("API_URL");
+const patientId = getEnvVarOrFail("PATIENT_ID");
+const facilityId = getEnvVarOrFail("FACILITY_ID");
+
```
Comment (Goncharo) [inline packages/utils/src/document/document-query.ts:11]: why is this script here? is it a leftover?

File: packages/api-sdk/src/medical/client/metriport.ts [hunk 36]
```diff
@@ -429,10 +431,15 @@ export class MetriportMedicalApi {
    * @param facilityId The facility providing the NPI to support this operation (optional).
    *        If not provided and the patient has only one facility, that one will be used.
    *        If not provided and the patient has multiple facilities, an error will be thrown.
+   * @param req.body Optional metadata to be sent through Webhook.
```
Comment (Goncharo) [inline packages/api-sdk/src/medical/client/metriport.ts:434]: nit: lowercase `webhook` for consistency

File: docs/_snippets/webhook-metadata-post-example.mdx [hunk 37]
```diff
@@ -0,0 +1,8 @@
+```json
+{
+  "metadata": {
+     youCan: "putAny",
+     stringKeyValue: "pairsHere",
+  }
+}
```
Comment (Goncharo) [inline docs/_snippets/webhook-metadata-post-example.mdx:7]: why the overhead of `metadata.metadata`, can we not just have metadata be the key-value pair dictionary?

ie suggest unwrapping metadata to just an object unless there's good reason otherwise

File: packages/utils/src/document/document-query.ts (hunk 35 above)
Comment (jonahkaye) [reply packages/utils/src/document/document-query.ts:11]: testing script. That was what I was told to test that sdk worked.

File: docs/_snippets/webhook-metadata-post-example.mdx (hunk 37 above)
Comment (leite08) [reply docs/_snippets/webhook-metadata-post-example.mdx:7]: I agree we don't need `metadata.metadata`. Just to be clear, I think this should be the payload/body to the API:

```json
{
  "metadata": {
    "someprop": "with whatever value"
  }
}
```

when using the SDK, I think this should be the experience:

```ts
const myPayload: Record<string, string> = { someprop: "with whatever value" };
await medicalApi.startDocumentQuery(patientId, facilityId, myPayload);
// ^ the SDK deals with wrapping the 3rd param into the `metadata` request body prop.
```

File: packages/api/src/routes/medical/schemas/request-metadata.ts (hunk 34 above)
[the remark this answers is not in the document]
Comment (leite08) [reply packages/api/src/routes/medical/schemas/request-metadata.ts:11]: Gotcha, I didn't see the `.optional()` in the definition of `medatada`; it makes sense

File: packages/api-sdk/src/medical/client/metriport.ts (hunk 32 above)
[the remark this answers is not in the document]
Comment (leite08) [reply packages/api-sdk/src/medical/client/metriport.ts:278]: The type of the param on the SDK does not mean its the same type/structure on the request body. There's no need for the user to send an object with 2 layers "enveloping" their data. See [this comment](https://github.com/metriport/metriport/pull/1087#discussion_r1384389282).

**Review by Goncharo — APPROVED** [review]
good to 🚢 pending the discussed SDK changes

File: packages/api-sdk/src/medical/client/metriport.ts [hunk 38]
```diff
@@ -264,16 +264,19 @@ export class MetriportMedicalApi {
    * @param dateFrom Optional start date that resources will be filtered by (inclusive). Format is YYYY-MM-DD.
    * @param dateTo Optional end date that resources will be filtered by (inclusive). Format is YYYY-MM-DD.
    * @param req.query.conversionType Optional to indicate how the medical record should be rendered.
+   * @param req.body Optional metadata to be sent through webhook
```
Comment (leite08) [inline packages/api-sdk/src/medical/client/metriport.ts:267]: This is SDK, there's no `req` param here. Can we update these docs (the new and the existing `req.query.conversionType` to represent the parameters, please?

Also, could add a bit more context to the param, maybe:

"Optional metadata to be sent along the webhook request as response of this query."

File: packages/api-sdk/src/medical/client/metriport.ts [hunk 39]
```diff
@@ -429,10 +432,16 @@ export class MetriportMedicalApi {
    * @param facilityId The facility providing the NPI to support this operation (optional).
    *        If not provided and the patient has only one facility, that one will be used.
    *        If not provided and the patient has multiple facilities, an error will be thrown.
+   * @param req.body Optional metadata to be sent through webhook.
```
Comment (leite08) [inline packages/api-sdk/src/medical/client/metriport.ts:435]: ditto: param name and description of the new one

File: packages/api/src/command/medical/patient/update-patient.ts (hunk 31 above)
[the remark this answers is not in the document]
Comment (leite08) [reply packages/api/src/command/medical/patient/update-patient.ts:36]: This doesn't need to be changed anymore, right?

File: docs/medical-api/more-info/webhooks.mdx [hunk 40]
```diff
@@ -12,13 +12,25 @@ This is the usual flow to obtain Patient data:
...
    - depending on the amount of data this can take some time to crunch together, so results are sent through a webhook message:
      [patient consolidated data](#patient-consolidated-data)
 
 To enable this integration approach with Metriport, and for some prerequesite reading to understand
 how the Webhook flow works, see [our Webhooks guide](/home/api-info/webhooks).
 
+### Passing Metadata
+
+You can pass metadata to the [start document query](/medical-api/api-reference/document/start-document-query) or [start consolidated data query](/medical-api/api-reference/fhir/consolidated-data-query-post), 
+endpoints, and you will receive the meta.data field of the webhook request.
```
Comment (leite08) [inline docs/medical-api/more-info/webhooks.mdx:25]: nit: Intentionally not escaping escaping `meta.data` or just missed it?

```
`meta.data`
```

**Review by leite08 — APPROVED** [review]
LGTM! Just left one small comment about formatting on the docs.
